File: ehs_frontend/src/api/ra.ts
import { api } from './client'
import type { RecommendResponse } from './inspect'

export async function generateRAJson(projectName: string, description: string): Promise<RecommendResponse> {
  const form = new FormData()
  form.append('project_name', projectName)
  form.append('description', description)
  const { data } = await api.post('/api/v1/ra/generate', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
  })
  return data
}

export async function downloadRADocx(projectName: string, description: string): Promise<void> {
  const form = new FormData()
  form.append('project_name', projectName)
  form.append('description', description)

  const response = await api.post('/api/v1/ra/generate/docx', form, {
    headers: { 'Content-Type': 'multipart/form-data' },
    responseType: 'blob',
  })

  const url = URL.createObjectURL(new Blob([response.data], {
    type: 'application/vnd.openxmlformats-officedocument.wordprocessingml.document',
  }))
  const a = document.createElement('a')
  a.href = url
  a.download = (projectName.trim() || 'risk_assessment') + '.docx'
  a.click()
  URL.revokeObjectURL(url)
}

File: ehs_backend/app/api/v1/ra.py
"""Req2 · Description → RA JSON + Word document download."""
import io
import re
from typing import Annotated

from fastapi import APIRouter, Form, HTTPException
from fastapi.responses import StreamingResponse

from app.schemas.inspect import RARow, RecommendResponse
from app.services.ra.json_docx_writer import json_to_docx
from app.services.recommendation.ra_generator import generate_ra_json

router = APIRouter(prefix="/ra")

DOCX_MIME = "application/vnd.openxmlformats-officedocument.wordprocessingml.document"


def _safe_filename(name: str) -> str:
    return re.sub(r"[^\w\- ]", "", name).strip().replace(" ", "_") or "risk_assessment"


@router.post(
    "/generate",
    tags=["Req2 · Description → Word Doc RA"],
    response_model=RecommendResponse,
)
async def generate_json(
    project_name: Annotated[str, Form()] = "",
    description: Annotated[str, Form()] = "",
) -> RecommendResponse:
    """Project description → structured full RA JSON."""
    project_text = description.strip() or project_name.strip()
    if not project_text:
        raise HTTPException(status_code=422, detail="Provide a project description or project name.")

    resolved_name = project_name.strip() or project_text
    ra_dict = await generate_ra_json(project_text)

    return RecommendResponse(
        activity=resolved_name,
        from_db=False,
        project=ra_dict.get("project", resolved_name),
        rows=[RARow(**r) for r in ra_dict.get("rows", [])],
        full_ra=ra_dict,
    )


@router.post(
    "/generate/docx",
    tags=["Req2 · Description → Word Doc RA"],
)
async def generate_docx(
    project_name: Annotated[str, Form()] = "",
    description: Annotated[str, Form()] = "",
) -> StreamingResponse:
    """Project description → generate full RA → download as .docx."""
    project_text = description.strip() or project_name.strip()
    if not project_text:
        raise HTTPException(status_code=422, detail="Provide a project description or project name.")

    resolved_name = project_name.strip() or "Untitled Project"
    ra_dict = await generate_ra_json(project_text)
    docx_bytes = json_to_docx(ra_dict, resolved_name)

    filename = _safe_filename(resolved_name) + ".docx"
    return StreamingResponse(
        io.BytesIO(docx_bytes),
        media_type=DOCX_MIME,
        headers={"Content-Disposition": f'attachment; filename="{filename}"'},
    )
